### PR TITLE
Fix InvalidCastException

### LINKED ISSUE #49148: Compile error combining ??= operator, local async functions, generic type parameter, and nullable property
**Version Used**: Visual Studio 16.8.0 Preview 6.0

**Steps to Reproduce**:

Compile this code:
```csharp
class Program
{
    static async Task Main(string[] args)
    {
        await Bar(new Baz());

        async Task Bar<T>(T baz) where T : Baz
        {
            baz.X ??= await Foo();
            async Task<int> Foo() => 0;
        }
    }

    class Baz
    {
        public int? X { get; set; }
    }
}
```

**Expected Behavior**:
Code should compile as normal.

**Actual Behavior**:
MSB6006 error with description '"csc.exe" exited with code -2146232797.'

The error specifically seems to be caused in a situation where there is an object passed in as a method parameter to an async generic local function, and whose type is of the generic type parameter constrained as a class; and then a nullable property of this object is then assigned the result of a nested async local function using the null-coalescing assignment operator.

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #49148

I'll add test soon.

### CODE REVIEW

**Comment by Youssef1313** [comment]
Since this bug involves multiple features together, I'm not sure what's the correct place to test this scenario. Can someone guide me?

File: src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/MethodToStateMachineRewriter.cs [hunk 1]
```diff
@@ -775,7 +775,7 @@ public override BoundNode VisitAssignmentOperator(BoundAssignmentOperator node)
 
             // We have an assignment to a variable that has not yet been assigned a proxy.
             // So we assign the proxy before translating the assignment.
-            return HoistRefInitialization((SynthesizedLocal)leftLocal, node);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/MethodToStateMachineRewriter.cs:778]: >(SynthesizedLocal) [](start = 42, length = 18)

Why is this the right change to make, given the assert for ```leftLocal.SynthesizedKind``` above?

**Comment by AlekseyTs** [comment]
@Youssef1313 Please provide more details about the nature of the issue. What is the local symbol we are dealing with? How did it came into existence? Why is it not synthesized even though it has ```SynthesizedLocalKind.Spill```? What scenarios are affected? Etc.

**Comment by Youssef1313** [comment]
I thought this should have been similar to what I done in #48260.
But that doesn't seem to be the case based on your feedback.
It needs a more deeper look. As I currently don't have access to a machine that can run Roslyn solution, I won't be able to try taking a deeper look. I'll close. Hoping that your comments will help who will work on this next.

**Comment by AlekseyTs** [comment]
> I thought this should have been similar to what I done in #48260.

That one was clear because ExceptionFilterPrologueOpt is statically typed as ```BoundStatementList``` and it was converted to ```BoundBlock```. The cast would never succeed for a non-null value.
I am assuming the code that is touched in this PR is working for majority of cases. It is quite possible that the proposed change is the right thing to do, but I think we need to get a better understanding what is going on and why.

**Comment by Youssef1313** [comment]
Yeah I agree that "why" is important. If I could get to a better machine or a VM soon I'll try to take a closer look. But I prefer to keep it closed to give a chance for anyone else who is interested to take a look.

**Pull request closed without merging**